Add documentation for BackendTLSPolicy (#2493)
kubernetes-sigs/gateway-api: site-src/guides/tls.md

Depending on the Listener Protocol, different TLS modes and Route types are supported.

> | Listener Protocol | TLS Mode    | Connection/s Affected   | Route Type Supported |

Reviewer: ```suggestion
| Listener Protocol | TLS Mode    | Connections Affected   | Route Type Supported |
```

|-------------------|-------------|-------------------------|----------------------|
| TLS               | Passthrough | Downstream and Upstream | TLSRoute             |
| TLS               | Terminate   | Downstream              | TCPRoute             |
| HTTPS             | Terminate   | Downstream              | HTTPRoute            |

> | HTTPS             | Backend     | Upstream                | HTTPRoute            |

Reviewer: I think the concepts of "downstream" and "upstream" are going to be very confusing here. I'd recommend copying the diagram you made for the GEP and adding it here along with the corresponding terminology. https://gateway-api.sigs.k8s.io/geps/gep-1897/#overview-what-do-we-want-to-do

Author: @robscott the concepts of downstream and upstream pre-exist in this document and the diagram.  I added a section called "Upstream", but there was already a section called "Downstream".    I can change "upstream" to "backend" here, but then what do I call what was formerly called "upstream" and not referencing the changes this introduces?

Author: Related comment, if I change "upstream" to "backend", shouldn't i also change "downstream" to "frontend'? Mixing the metaphors "downstream" and "backend" doesn't seem less confusing.

Reviewer: 🤦 you're right, I was not looking at the full deploy preview and related context, in which case the existing terminology you've used makes sense. 

I do really like the diagram you created for the GEP though, so if you wanted to make a version that uses the terminology in this doc (upstream and downstream), that could be a great upgrade on the current diagram here. (That's a non-blocking/optional suggestion though).

On further inspection, I would actually revert the change to this table. While "Passthrough" and "Terminate" are options on Listener TLS, "Backend" is not and is somewhat orthogonal.

Please note that in case of `Passthrough` TLS mode, no TLS settings take
effect as the TLS session from the client is NOT terminated at the Gateway, but rather
passes through the Gateway, encrypted.
In the case of `Backend`, only the Upstream connection is affected.
For `HTTPRoute`, the use of both `Terminate` and `Backend` is supported, and provides what

> is commonly known as a connection that is terminated and then re-encrypted at the Gateway.

Reviewer: Similar comment as above, I think it's best to revert this change since `Backend` is not really the same thing as `Terminate` or `Passthrough`.

Author: Done.

{% include 'standard/tls-cert-cross-namespace.yaml' %}
```

## Upstream TLS

Upstream TLS settings are configured using the experimental `BackendTLSPolicy`

> applied to a `Service` that accesses the backend to be targeted.

Reviewer: ```suggestion
attached to a `Service` via a target reference.
```

> This type of TLS is used when the backend pod has its own certificate and the
> application developer needs to convey how to successfully connect to the
> backend pod from the Gateway.

Reviewer: ```suggestion

This resource can be used to describe the SNI the Gateway should use to connect to the
backend and how the certificate served by the backend Pod(s) should be verified.
```

### TargetRefs and TLS

The `TargetRef` identifies the `Service` for which your HTTPRoute requires TLS.
The `TLS` configuration for upstream TLS contains a required `Hostname`, and either
`CACertRefs` or `WellKnownCACerts`.
`Hostname` refers to the SNI the Gateway should use to connect to the backend, and
must match the certificate served by the backend pod.
`CACertRefs` refer to one or more PEM-encoded TLS certificates.
If there are no specific certificates to use, then you must set `WellKnownCACerts` to
"System" to tell the Gateway to use a set of trusted CA Certificates. There may be some
variation in which system certificates are used by each implementation. Refer to documentation

> from your implementation of choice for more information.

Reviewer: Would recommend adding some newlines here to split this into more than one paragraph. The deploy preview merges this all into one paragraph that can be a bit hard to follow: https://deploy-preview-2493--kubernetes-sigs-gateway-api.netlify.app/guides/tls/#targetrefs-and-tls

!!! info "Restrictions"

    - Cross-namespace certificate references are not allowed.
    - Wildcard hostnames are not allowed.

### Examples

#### Using System Certificates

In this example, the `BackendTLSPolicy` is configured to use system certificates
to connect with a TLS-encrypted upstream connection where Pods backing the
`dev` Service are expected to serve a valid certificate for `dev.example.com`.

```yaml
{% include 'experimental/v1alpha2/backendtlspolicy-system-certs.yaml' %}
```

#### Using Explicit CA Certificates

In this example, the `BackendTLSPolicy` is configured to use certificates defined
in the configuration map `auth-cert`
to connect with a TLS-encrypted upstream connection where Pods backing the

> `auth` Service are expected to serve a valid certificate for `auth.example.com`.

Reviewer: Tiny nit: Line length is a bit inconsistent here.

{% include 'standard/tls-cert-cross-namespace.yaml' %}
```

## Upstream TLS

Upstream TLS settings are configured using the experimental `BackendTLSPolicy`
applied to a `Service` that accesses the backend to be targeted.
This type of TLS is used when the backend pod has its own certificate and the
application developer needs to convey how to successfully connect to the
backend pod from the Gateway.

### TargetRefs and TLS

The `TargetRef` identifies the `Service` for which your HTTPRoute requires TLS.
The `TLS` configuration for upstream TLS contains a required `Hostname`, and either
`CACertRefs` or `WellKnownCACerts`.
`Hostname` refers to the SNI the Gateway should use to connect to the backend, and
must match the certificate served by the backend pod.
`CACertRefs` refer to one or more PEM-encoded TLS certificates.
If there are no specific certificates to use, then you must set `WellKnownCACerts` to

> "System" to tell the Gateway to use a set of trusted CA Certificates.  Refer to your
> implementation documentation for more details.

Reviewer: ```suggestion
"System" to tell the Gateway to use a set of trusted CA Certificates. There may be some
variation in which system certificates are used by each implementation. Refer to documentation
from your implementation of choice for more information.
```

!!! info "Restrictions"

    - Cross-namespace certificate references are not allowed.
    - Wildcard hostnames are not allowed.

### Examples

#### Using System Certificates

In this example, the `BackendTLSPolicy` is configured to use system certificates

> to connect with a TLS-encrypted upstream connection to the pod called `dev-pod`,
> with hostname `dev-pod.example.com`, and
> served by the service called `dev-service`.

Reviewer: Similar to above, would avoid suggesting that this refers to a single Service.
```suggestion
to connect with a TLS-encrypted upstream connection where Pods backing the
`dev` Service are expected to serve a valid certificate for `dev.example.com`.
```

Author: I think you meant "to avoid suggesting that this refers to a single Pod"?  Agreed.

```yaml
{% include 'experimental/v1alpha2/backendtlspolicy-system-certs.yaml' %}
```

#### Using Explicit CA Certificates

In this example, the `BackendTLSPolicy` is configured to use certificates defined
in the configuration map `auth-cert`
to connect with a TLS-encrypted upstream connection to the pod called `auth-pod`,
with hostname `auth-pod.example.com`, and

> served by the service called `auth-service`.

Reviewer: Would recommend a similar set of changes for this example.
